File: 94_Binary_Tree_Inorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> values;
        return inorderTraversal(root, values);
    }
    
    vector<int> inorderTraversal(TreeNode* root, vector<int> values)
    {
        if(root == NULL) return values; 
        if(root->left != NULL) values = inorderTraversal(root->left, values); 
        values.push_back(root->val); 
        if(root->right != NULL) values = inorderTraversal(root->right, values);
        return values; 
    }
};

File: 94_Binary_Tree_Inorder_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> r = new ArrayList<Integer>(); 
        return inorderTraversal(root, r); 
    }
    
    public List<Integer> inorderTraversal(TreeNode root, List<Integer> r)
    {
        if(root == null) return r; 
        if(root.left != null) r = inorderTraversal(root.left, r); 
        r.add(root.val); 
        if(root.right != null) r = inorderTraversal(root.right, r); 
        return r; 
    }
}